design/StableFunctions.md

```diff
@@ -0,0 +1,126 @@
+
+Note: This is not supported with classical persistence.
+
+With this, the only non-persistent (non-stable) values remain to be lambdas, async handles (aka continuations, futures), and local async function references.
+
+## Example
+
+With this, the following program can now be used for orthogonal persistence. For this example, you need a [specific base library branch](https://github.com/dfinity/motoko-base/tree/luc%2Fstable-functions), which contains slight modifications for supporting stable functions.
+
+```
+import HashMap "base/HashMap";
+import Hash "base/Hash";
+import Nat "base/Nat";
+import Iter "base/Iter";
+
+persistent actor {
+    let hashMap = HashMap.HashMap<Nat, Text>(10, Nat.equal, Hash.hash); // finally persistent!
+
+    public func populate(): async() {
+        hashMap.put(1, "A");
+        hashMap.put(2, "B");
+        hashMap.put(3, "C");
+    };
+
+    public func list(): async [(Nat, Text)] {
+        Iter.toArray<(Nat, Text)>(hashMap.entries());
+    };
+}
+```
+
+## Stable Functions and Stable Scopes
+
+A stable function is a named non-async local function in a stable scope, only closing over variables of a stable type.
+
+A stable scope is:
+* the main actor,
+* an actor class,
+* a module imported with a unique identifier from a stable scope,
+* a named non-async function in a stable scope,
+* a class in a stable scope, or,
+* a named object in a stable scope.
+
+Generic type parameters of stable functions and stable classes are bounded to stable types.
+
+A stable function is also a stable type.
+
+Syntactically, function types are prefixed by `stable` to denote a stable function, e.g.
+`stable X -> Y`. Stable functions implicitly have a corresponding stable reference type.
+
+A stable function type is a sub-type of a flexible function type with type-compatible signature, i.e. `stable X' -> Y <: X -> Y'` for `X' <: X` and `Y' :< Y`.
+
+## Upgrades of Stable Functions
+
+Stable functions are upgraded as follows:
+* All stable functions that are reachable from stable variables are considered alive.
+* Each alive stable function must have a matching declaration in the new program version.
+* Stable functions match between program versions if they have an equal fully qualified name.
+* For matching functions, the function type of the new version must be compatible to the previous version (super-type).
+* For matching functions, the closure type in the new version must be compatible with the previous version, see below.
+
+All other functions, such as lambdas, named functions in a lambda, async functions, or functions imported from a module without a unique import identifier, are flexible functions.
+
+## Stable Closures
+
+The closures of stable functions are represented in a portable format with the following clearly defined binding of captured variables:
+* Local variables are captured by name.
+* Parameters are captured by position (also considering nested functions).
```

Comment:
I'm not sure if position is really well defined since we have argument binding by pattern matching.